### PR TITLE
Improve IHS Data All merge process

### PR DESCRIPTION
**Pull request by crlang44**
IHS merging process now corrects any overlapping ownership entries using a priority table.
Add marine exchange integration using similar merging process

### CODE REVIEW

File: functions/update-ihs-data-all/marine_exchange.sql [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        mapping 
+      WHERE 
+        ihs.operator = mapping.operator_ihs 
+        AND ihs.mmsi = mapping.mmsi 
+        AND (
+          ihs.start_date < mapping.start_date 
+          AND ihs.end_date > mapping.start_date 
+          AND ihs.end_date <= mapping.end_date
+        )
+    ) cte
```
Comment (crlang44) [inline functions/update-ihs-data-all/marine_exchange.sql:174]: Terrible variable name, please change

File: functions/update-ihs-data-all/update_tables.sql [hunk 2]
```diff
@@ -0,0 +1,215 @@
+-- Create a new table that combines continuous segments of ownership from the new table for comparing against the original table segments
```
Comment (crlang44) [inline functions/update-ihs-data-all/update_tables.sql:1]: Would be good to add comments using the period relationships used in the marine_exchange.sql for consistency

File: sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ais_segments_agg.sql [hunk 3]
```diff
@@ -195,9 +195,11 @@ FROM
                       CASE WHEN final_speed_bin_num = LEAD(final_speed_bin_num) OVER w THEN 0 WHEN final_speed_bin_num <> LEAD(final_speed_bin_num) OVER w THEN 1 END AS seg_chg, 
                     FROM 
                       `benioff-ocean-initiative.whalesafe_v3.ais_segments`  AS s
-                      LEFT JOIN `whalesafe_v3.vsr_zones` AS z
+                      FULL OUTER JOIN `whalesafe_v3.vsr_zones` AS z
```
Comment (crlang44) [inline sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ais_segments_agg.sql:198]: Why did we switch to use outer join here?

File: sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ais_segments_agg.sql [hunk 4]
```diff
@@ -195,9 +195,11 @@ FROM
...
                       ON ST_Intersects(s.linestring, z.geog)
                     WHERE (timestamp) > new_seg_agg_ts
+                    AND
```
Comment (crlang44) [inline sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ais_segments_agg.sql:201]: Why did we add this?

File: sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ship_stats_daily.sql [hunk 5]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< HEAD
```
Comment (crlang44) [inline sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ship_stats_daily.sql:1]: Need to remove this

File: sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ship_stats_monthly.sql [hunk 6]
```diff
@@ -1,309 +1,310 @@
...
-        MAX(max_timestamp)
-      ) 
-    FROM 
-      `whalsafe_v3.ship_stats_monthly` 
-    LIMIT 
-      1
-  ) AS newest_date, 
-  CURRENT_TIMESTAMP() AS date_accessed, 
-  'ship_stats_monthly' AS table_name;
+<<<<<<< HEAD
```
Comment (crlang44) [inline sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ship_stats_monthly.sql:1]: Need to remove this

File: sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ais_segments_agg.sql (hunk 4 above)
Comment (crlang44) [reply sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ais_segments_agg.sql:201]: Merge mistake. Manually removed this again

File: sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ais_segments_agg.sql (hunk 3 above)
Comment (crlang44) [reply sql_v3_scheduled-on-BigQuery/update_whalesafe_v3_ais_segments_agg.sql:198]: Merge mistake. Manually changed back to LEFT. As documented by previous commit. BigQuery does not support FULL JOIN without equality comparison

File: functions/update-ihs-data-all/main.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+from google.cloud import bigquery
+from google.oauth2 import service_account
+
+import functions_framework
+
+# CloudEvent function to be triggered by an Eventarc Cloud Audit Logging trigger
+# Note: this is NOT designed for second-party (Cloud Audit Logs -> Pub/Sub) triggers!
+
+EVALUATE = True
+SAVE_IHS = False
```
Comment (crlang44) [inline functions/update-ihs-data-all/main.py:13]: Update this to True when ready or make it an environment variable to the cloud function

File: functions/update-ihs-data-all/marine_exchange.sql [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+          AND mapping.end_date >= ihs.end_date
+        )
+    ) corrected_ihs_data 
+  WHERE 
+    ihs_temp_table.mmsi = corrected_ihs_data.mmsi 
+    AND (
+      ihs_temp_table.start_date = corrected_ihs_data.start_date 
+      AND ihs_temp_table.end_date = corrected_ihs_data.end_date
+    );
+  --      Enclosing
```
Comment (calliesteffen) [inline functions/update-ihs-data-all/marine_exchange.sql:262]: This doesnt look right -

File: functions/update-ihs-data-all/marine_exchange.sql [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+        mapping.mmsi as mmsi, 
+        mapping.start_date as start_date, 
+        mapping.end_date as end_date 
+      FROM 
+        {mapping_table} mapping
+    ) 
+    SELECT 
+      ihs.* REPLACE(
+        mapping.end_date AS end_date,
+        mapping.operator_update_to AS operator,
```
Comment (calliesteffen) [inline functions/update-ihs-data-all/marine_exchange.sql:286]: need to update the end date to add +1 day

File: functions/update-ihs-data-all/marine_exchange.sql (hunk 8 above)
Comment (crlang44) [reply functions/update-ihs-data-all/marine_exchange.sql]: It was correct but I updated the naming and the ordering to make more sense

File: functions/update-ihs-data-all/marine_exchange.sql (hunk 9 above)
Comment (crlang44) [reply functions/update-ihs-data-all/marine_exchange.sql]: Updated

File: functions/update-ihs-data-all/update_tables.sql (hunk 2 above)
Comment (crlang44) [reply functions/update-ihs-data-all/update_tables.sql]: Added

**Pull request merged**